nvflare/apis/utils/reliable_message.py

```diff
@@ -0,0 +1,369 @@
+
+            time.sleep(2.0)
+
+    @classmethod
+    def shutdown(cls):
+        cls._executor.shutdown(cancel_futures=True, wait=False)
+        cls._shutdown_asked = True
+
+    @classmethod
+    def send_request(
+        cls, target: str, topic: str, request: Shareable, timeout: float, abort_signal: Signal, fl_ctx: FLContext
+    ) -> Shareable:
+        tx_id = str(uuid.uuid4())
+        receiver = _ReplyReceiver(tx_id)
+        cls._reply_receivers[tx_id] = receiver
+        request.set_header(HEADER_TX, tx_id)
+        request.set_header(HEADER_OP, OP_REQUEST)
+        request.set_header(HEADER_TOPIC, topic)
+        request.set_header(HEADER_TIMEOUT, timeout)
+        try:
+            result = cls._send_request(target, request, timeout, abort_signal, fl_ctx, receiver)
+        except Exception as e:
+            result = _error_reply(ReturnCode.ERROR, str(e))
+        cls._reply_receivers.pop(tx_id)
+        return result
+
+    @classmethod
+    def _send_request(
+        cls,
+        target: str,
+        request: Shareable,
+        timeout: float,
+        abort_signal: Signal,
+        fl_ctx: FLContext,
+        receiver: _ReplyReceiver,
+    ) -> Shareable:
+        engine = fl_ctx.get_engine()
+
+        # keep sending the request until a positive ack or result is received
+        num_tries = 0
+        while True:
+            if abort_signal.triggered:
+                return make_reply(ReturnCode.TASK_ABORTED)
+
+            ack = engine.send_aux_request(
+                targets=[target],
+                topic=TOPIC_RELIABLE_REQUEST,
+                request=request,
+                timeout=timeout,
+                fl_ctx=fl_ctx,
+            )
+            ack, rc = _extract_result(ack, target)
+            if ack and rc != ReturnCode.COMMUNICATION_ERROR:
+                # is this result?
+                op = ack.get_header(HEADER_OP)
+                if op == OP_REPLY:
+                    # the reply is already the result - we are done!
+                    # this could happen when we didn't get positive ack for our first request, and the result was
+                    # already produced when we did the 2nd request (this request).
+                    return ack
+
+                # the ack is a status report - check status
+                status = ack.get_header(HEADER_STATUS)
+                if status and status != STATUS_NOT_RECEIVED:
+                    # status should never be STATUS_NOT_RECEIVED, unless there is a bug in the receiving logic
+                    # STATUS_NOT_RECEIVED is only possible during "query" phase.
+                    break
+
+            # we didn't get a positive ack - wait a short time and re-send the request.
+            num_tries += 1
+            if num_tries > cls._max_retries:
+                # enough tries
+                return _error_reply(ReturnCode.COMMUNICATION_ERROR, f"Max send retries ({cls._max_retries}) reached")
+            time.sleep(cls._query_interval)
```

Comment:
should this be "resend_interval" or a fast checking period, then we use other "if" statement to determine if we need to resend?

Reply:
For example:

'''
While True:

    If current time - previous time > resend interval:
         Do resend 

    time.sleep(0.1)
'''

Reply:
Changed to a fast loop.